### ISSUE TITLE
Electrum 4.1.2 is changing my manual fees

### ISSUE THREAD

**Issue by santimaster2000** [spesmilo/electrum#7216]
I don't know how to explain this, after "updating" (I just downloaded a new portable .exe) my Electrum wallet, the thing just refuses to use the exact fee I want, for example, right now, wanted to send a small payment, choose 20 sats, went pay, advanced then static, then I wrote 20 in the fee "box", finalize, sign, broadcast, everything alright ? I go to check on my transaction history, the payment went through but instead of 20 sats it took 20.1, yeah yeah, I know, it's not that much money TODAY, but I thought that was the point, we believe that Bitcoin will be incredibly valuable in the future, that's why we hold, at least that's why I hold, and I don't like my wallet changing my fees.

Have tried unchecking "enable output value rounding", the issue persists, have also tried typing 20.0 instead of 20, the issue persists, it shows the right amount before sending the tx, but afterwards when I check on my tx history, it shows that it added .2, .3, .4 sats sometimes.

This is my first time reporting a bug, sorry if formating is not appropriate.

**Comment by SomberNight** [comment]
It's hard to debug this without a concrete example. That is, if you can reproduce, please upload a screenshot of the transaction details window before clicking "finalize", and then another after having signed the transaction.

In general, I would say that transaction size estimation (and hence fee estimation) always has some wiggle-room.
For example, the size of the signature is not known in advance, so we err on the side of overestimating it, which means the feerate the user enters is actually a lower bound.

> Have tried unchecking "enable output value rounding", the issue persists

Right, that's another thing that can affect this.

Also, probably not the cause in this case, but another point is that we are dropping the change output if it would be dust, which increases the fee in that case quite a bit.

**Comment by santimaster2000** [comment]
I can reproduce it, sadly, I don't like the idea of showing you (Or anyone else), my txs, or my wallet info.

This is more of a "Hey, this is broken, it wasn't before v4, you guys might wanna look into it" not a "I need this fixed".

What I'm saying is, I can live with the bug, but it is a bug.

**Comment by SomberNight** [comment]
You can try on testnet and screenshot there.

> What I'm saying is, I can live with the bug, but it is a bug.

It might not be a bug.
Like I said, an example is the size of the signature not being known in advance.

**Comment by santimaster2000** [comment]
Could you give me a basic guide on what exactly would I have to do in order to show you ? I've never used testnet.

I'm sorry, I don't know what the size of the signature is.

**Comment by SomberNight** [comment]
What does the window show regarding the size of the transaction (in bytes) before you click "finalize", and what does it show after signing it?

Also what fee are you setting at the bottom of the window; and what fee does it show near the top of the window?
-------- Original Message --------

**Comment by santimaster2000** [comment]
I'll check all of those things when I send a new tx, would that be OK ?

**Comment by Milo007-github** [comment]
> Could you give me a basic guide on what exactly would I have to do in order to show you ? I've never used testnet.

Search for "Electrum Testnet" in the windows search bar. The testnet application has "Electrum Testnet" written on its icon just like "Electrum" is written on the icon of the main application.

**Comment by santimaster2000** [comment]
> 
> 
> > Could you give me a basic guide on what exactly would I have to do in order to show you ? I've never used testnet.
> 
> Search for "Electrum Testnet" in the windows search bar. The testnet application has "Electrum Testnet" written on its icon just like "Electrum" is written on the icon of the main application.

Read my original post, I'm running the portable version of Electrum.

**Comment by Milo007-github** [comment]
> Read my original post, I'm running the portable version of Electrum.

Do you know others who is running an installed version and facing the same issue? 

It could be that the problem is specific to the portable version.

**Comment by santimaster2000** [comment]
> 
> 
> > Read my original post, I'm running the portable version of Electrum.
> 
> Do you know others who is running an installed version and facing the same issue?
> 
> It could be that the problem is specific to the portable version.

Not really.

I can try with the .AppImage version next time.

**Comment by romanalien** [comment]
I have the same problem, my Electrum changes fees too. Iam sending 3 screens with errors.
![01 Error ETA](https://user-images.githubusercontent.com/38290773/115156027-b6eb5200-a082-11eb-85c7-c2c3bffc5785.jpg)
![02 change static](https://user-images.githubusercontent.com/38290773/115156028-b783e880-a082-11eb-8609-674086e819e9.jpg)
![03-Error Target fee](https://user-images.githubusercontent.com/38290773/115156029-b8b51580-a082-11eb-8179-1467b4b153a9.jpg)

**Comment by ecdsa** [comment]
> I have the same problem, my Electrum changes fees too. Iam sending 3 screens with errors.

No it is not. You are mistaking the change amount and the fee.
The fee is on the 3rd screenshot is exactly the same as on the second screenshot.
Please learn what "change" is in the context of bitcoin.

**Comment by romanalien** [comment]
Target fee is 0,0001445 on 3rd screenshot, but payment is 0,037971 (yellow btc adresse).
 The payment server ends for high fee.
Please help me, thank you.

**Comment by romanalien** [comment]
The 3rd screenshot
I send 0.01 BTC
second unknown address ??? (yellow) about The fee 0.037971BTC

**Comment by Milo007-github** [comment]
> > I have the same problem, my Electrum changes fees too. Iam sending 3 screens with errors.
> 
> No it is not. You are mistaking the change amount and the fee.
> The fee is on the 3rd screenshot is exactly the same as on the second screenshot.
> Please learn what "change" is in the context of bitcoin.

![115156029-b8b51580-a082-11eb-8179-1467b4b153a9](https://user-images.githubusercontent.com/82448112/115160699-43b00300-a0b7-11eb-8a08-024f181ddafb.jpg)


You missed out on the disparity in the total fee being shown at the bottom and the top. On the top right hand side besides the "Fee :" it's written 0.00014522 BTC while at the bottom right hand side the target fee (5 sat/byte) multiplied by the transaction size (2890 bytes) results in a total fee of 0.0001445 BTC. There's a difference of 72 satoshis. How do you explain that?

**Comment by Milo007-github** [comment]
> The 3rd screenshot
> I send 0.01 BTC
> second unknown address ??? (yellow) about The fee 0.037971BTC

Please go through the link to know more about "change addresses"

https://bitzuma.com/posts/who-needs-bitcoin-change-addresses-anyway/

The second address highlighted in yellow is a change address (one of the many addresses in your own wallet where you recieve a "change" balance of bitcoin as a return). It's not the fee of the transaction as you think. You could see 19 addresses (all your wallet addresses with bitcoins in them) as inputs in your transaction details.  The cumulative balance of all these inputs addresses minus the sum of the amount you wish to send (0.01) and the transaction fee (0.00014522) is what you recieve back in one of your wallet address which is then called a "change address".

**Comment by SomberNight** [comment]
> You missed out on the disparity in the total fee being shown at the bottom and the top. On the top right hand side besides the "Fee :" it's written 0.00014522 BTC while at the bottom right hand side the target fee (5 sat/byte) multiplied by the transaction size (2890 bytes) results in a total fee of 0.0001445 BTC. There's a difference of 72 satoshis. How do you explain that?

There is a blue information icon right next to the red circle the user drew for the sole purpose of explaining the difference...
Looking at the screenshot and considering the difference is only 22 sats, it is most likely due to "output value rounding" being enabled.

**Comment by romanalien** [comment]
Well, why can't I pay?


![04-payment](https://user-images.githubusercontent.com/38290773/115199698-c48cf000-a0f3-11eb-8bf5-c3dd70249922.jpg)

**Comment by Milo007-github** [comment]
> Well, why can't I pay?
> 
> ![04-payment](https://user-images.githubusercontent.com/38290773/115199698-c48cf000-a0f3-11eb-8bf5-c3dd70249922.jpg)

Because 5.1 sat/byte is an incredibly low fee rate at the moment and the error message you are getting is self explanatory. Try setting a higher fee. You can update yourself with the current market situation in mempool.space.

**Comment by Milo007-github** [comment]
> There is a blue information icon right next to the red circle the user drew for the sole purpose of explaining the difference...
> Looking at the screenshot and considering the difference is only 22 sats, it is most likely due to "output value rounding" being enabled.

Indeed correct.

With output value rounding disabled :

Before finalizing...

![Screenshot 2021-04-19 134552](https://user-images.githubusercontent.com/82448112/115207175-f8c6d780-a118-11eb-8cd7-23e3f743ea6b.png)

After finalizing...

![Screenshot 2021-04-19 134846](https://user-images.githubusercontent.com/82448112/115207298-172cd300-a119-11eb-8f9c-7c74cee09793.png)

After signing...

![Screenshot 2021-04-19 134935](https://user-images.githubusercontent.com/82448112/115207366-2744b280-a119-11eb-8327-65daaddce301.png)


With output value rounding enabled :

Before finalizing...

![Screenshot 2021-04-19 140512](https://user-images.githubusercontent.com/82448112/115207850-97ebcf00-a119-11eb-8ecb-825ca6030a1f.png)

![Screenshot 2021-04-19 141720](https://user-images.githubusercontent.com/82448112/115208292-06c92800-a11a-11eb-826e-a841d2c10572.png)

After finalizing...

![Screenshot 2021-04-19 140538](https://user-images.githubusercontent.com/82448112/115208351-15174400-a11a-11eb-87ec-cb643d162ed3.png)

After signing...

![Screenshot 2021-04-19 140636](https://user-images.githubusercontent.com/82448112/115208457-34ae6c80-a11a-11eb-9bf3-9ce15e2c6309.png)

Note: Even though the transaction size reduces by 1 byte in both cases after signing the total fee remains unchanged before and after signing.

**Comment by ecdsa** [comment]
> You missed out on the disparity in the total fee being shown at the bottom and the top. On the top right hand side besides the "Fee :" it's written 0.00014522 BTC while at the bottom right hand side the target fee (5 sat/byte) multiplied by the transaction size (2890 bytes) results in a total fee of 0.0001445 BTC. There's a difference of 72 satoshis. How do you explain that?

No, I did not miss that. That user simply is confusing change and fee, and my answer to him was on point.

The fee shown at the bottom is the "target fee", the fee at the top is the actual fee (after rounding). 
If these values were supposed to be the same, we would not display the same information twice.

**Comment by Milo007-github** [comment]
> No, I did not miss that. That user simply is confusing change and fee, and my answer to him was on point.
> 
> The fee shown at the bottom is the "target fee", the fee at the top is the actual fee (after rounding).
> If these values were supposed to be the same, we would not display the same information twice.

Got it. So if output value rounding is disabled the "target fee" and "actual fee" becomes exactly the same as I can produce.

**Comment by santimaster2000** [comment]
OK, I'm back, I checked, the fee changes after I click the "Sign" button, also, I'm not using a change address (As far as I understand), I'm working with 1 public key, 1 private key, only 1 address, the fee is still changing after I sign the tx.

This didn't happen before version 4, so unless something change about how Electrum displays the final fee, then it would seem there's a problem.

**Comment by SomberNight** [comment]
Is it the absolute fee that changed or the feerate (fee per byte)? By how much?

**Comment by santimaster2000** [comment]
> 
> 
> Is it the absolute fee that changed or the feerate (fee per byte)? By how much?

Now that you mention it, it's only the feerate, the absolute fee seems to stay the same, weird.

**Comment by SomberNight** [comment]
Ok, that is not a bug. Like I said in my first reply,

> In general, I would say that transaction size estimation (and hence fee estimation) always has some wiggle-room.
For example, the size of the signature is not known in advance, so we err on the side of overestimating it, which means the feerate the user enters is actually a lower bound.

Every input you spend needs to be signed to prove ownership. For a standard wallet, each input will contain one signature. The signature commits to most of the data in the transaction, so by the time the signature is being created everything needs to be set. This includes the absolute fee the transaction pays.
There is an added complication that when wanting to set the fee rate (fee per byte) instead of the absolute fee: while the absolute fee can be ~freely set, the size of the transaction depends on a few things. In particular, the signatures in the inputs are also part of the final transaction, so their size also affects the size of the tx, and [their size is variable](https://bitcoin.stackexchange.com/a/12556). In practice, DER-encoded ECDSA signatures (including the sighash byte) are either 71 or 72 bytes (note that in case of segwit, they are discounted by -75%).

In short, we cannot know the precise size of the transaction before signing the tx, but by then the absolute fee is already fixed, which means it is in general not possible to set a precise feerate for a transaction. The code is pessimistic and always overestimates the tx size, i.e. sometimes you end up paying a higher feerate than set (but never a lower one).
